File: Day5/Problem_15.py
"""  Topic :- Maths
Problem Name :- Implement Pow(x,n) | X raised to the power N
Problem Statement :- Implement pow(x, n), which calculates x raised to the power n (i.e., x^n).

(Time Complexity :- O(log n) and Space Complexity :- O(1))

"""


"""def myPow(x, n):
    ans = 1
    for _ in range(n):
        ans *= x
    return ans

"""


def myPow(x, n):
    ans = 1.0
    power = abs(n)

    while power > 0:
        if power % 2 == 1:
            ans *= x
            power -= 1
        else:
            x *= x
            power //= 2

    if n < 0:
        return 1 / ans
    
    return ans


if __name__ == "__main__":
    x = float(input())
    n = int(input())

    print(myPow(x, n))

File: Day5/Problem_15.cpp
/* Topic :- Maths
Problem Name :- Implement Pow(x,n) | X raised to the power N
Problem Statement :- Implement pow(x, n), which calculates x raised to the power n (i.e., x^n).
*/

// TC : O(log n) and SC : O(1)



#include <bits/stdc++.h>
using namespace std;

double myPow(double x, int n) {
    long long power = n;
    
    if (power < 0)
        power = -power;

    double ans = 1.0;

    while (power > 0) {
        if (power % 2 == 1) {
            ans *= x;
            power--;
        } else {
            x *= x;
            power /= 2;
        }
    }

    if (n < 0)
        ans = 1.0 / ans;

    return ans;
}

int main() {
    double x;
    int n;

    cin >> x >> n;

    cout << myPow(x, n);

    return 0;
}